src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -743,169 +743,180 @@ HeapWord* ShenandoahFreeSet::allocate_single(ShenandoahAllocRequest& req, bool&
-              flip_to_old_gc(r);
-            } else {
-              flip_to_gc(r);
-            }
-            // Region r is entirely empty.  If try_allocat_in fails on region r, something else is really wrong.
-            // Don't bother to retry with other regions.
-            log_debug(gc, free)("Flipped region " SIZE_FORMAT " to gc for request: " PTR_FORMAT, idx, p2i(&req));
-            return try_allocate_in(r, req, in_new_region);
-          }
-          idx = _partitions.find_index_of_previous_available_region(ShenandoahFreeSetPartitionId::Mutator, idx - 1);
-        }
-      }
-      // No dice. Do not try to mix mutator and GC allocations, because adjusting region UWM
-      // due to GC allocations would expose unparsable mutator allocations.
-      break;
+  bool allow_new_region = can_allocate_in_new_region(req);
+  if (allow_new_region) {
+    // Try a free region that is dedicated to GC allocations.
+    result = allocate_from_partition_with_affiliation(req.is_old()? ShenandoahFreeSetPartitionId::OldCollector:
+                                                      ShenandoahFreeSetPartitionId::Collector,
+                                                      ShenandoahAffiliation::FREE, req, in_new_region);
+    if (result != nullptr) {
+      return result;
     }
+  }
+
+  // No dice. Can we borrow space from mutator view?
+  if (!ShenandoahEvacReserveOverflow) {
+    return nullptr;
+  }
+
+  if (!allow_new_region && req.is_old() && (_heap->young_generation()->free_unaffiliated_regions() > 0)) {
+    // This allows us to flip a mutator region to old_collector
+    allow_new_region = true;
+  }
+
+  // We should expand old-gen if this can prevent an old-gen evacuation failure.  We don't care so much about
+  // promotion failures since they can be mitigated in a subsequent GC pass.  Would be nice to know if this
+  // allocation request is for evacuation or promotion.  Individual threads limit their use of PLAB memory for
+  // promotions, so we already have an assurance that any additional memory set aside for old-gen will be used
+  // only for old-gen evacuations.
+  if (allow_new_region) {
+    // Try to steal an empty region from the mutator view.
+    result = try_allocate_from_mutator(req, in_new_region);
+  }
+
+  // This is it. Do not try to mix mutator and GC allocations, because adjusting region UWM
+  // due to GC allocations would expose unparsable mutator allocations.
+  return result;
+}
+
+bool ShenandoahFreeSet::can_allocate_in_new_region(const ShenandoahAllocRequest& req) {
+  if (!_heap->mode()->is_generational()) {
+    return true;
+  }
+
+  assert(req.is_old() || req.is_young(), "Should request affiliation");
+  return (req.is_old() && _heap->old_generation()->free_unaffiliated_regions() > 0)
+         || (req.is_young() && _heap->young_generation()->free_unaffiliated_regions() > 0);
+}
+
+HeapWord* ShenandoahFreeSet::try_allocate_from_mutator(ShenandoahAllocRequest& req, bool& in_new_region) {
```

Comment:
Why does this always use right-biased allocation from the mutator partition? (And not update allocation bias state?) Is this intentional? I assume it's because these aren't fresh allocations, but evacuations. Do evacuations with right bias end up closer to allocations for other evacuations? A brief comment explaining this difference from mutator allocation would be useful.

Reply:
I think it makes sense that the collector would take regions from the mutator according to the allocation bias for the _collector_. @kdnilsen , what do you think?

Reply:
When the collector is trying to allocate from the "mutator partition", it biases towards high end of memory because our general goal is long-lived data (except humongous objects which live at the low-end of memory) lives at high end of memory.  Survivor regions (regions allocated by the collector) are assumed to be longer lived than ephemeral objects (regions allocated by the mutator).

Reply:
Makes sense. A single line comment to that effect would be great.